common/src/main/java/com/proximyst/ban/service/MessageService.java

```diff
@@ -0,0 +1,230 @@
+        this.messageFactory.staticComponent(KEY_TARGET_NAME, target.getUsername()),
+        this.messageFactory.staticComponent(KEY_TARGET_UUID, target.getUuid().toString()),
+        this.messageFactory.staticComponent(KEY_QUANTITY, Integer.toString(amount)));
+  }
+
+  public @NonNull IMessage commandsHistoryEntry(final @NonNull Punishment punishment) {
+    return this.messageFactory.placeholderMessage(MessageKey.COMMANDS_HISTORY_ENTRY,
+        this.messageFactory.componentComponent(this.createPlaceholders(punishment)));
+  }
+  // </editor-fold>
+
+  public @NonNull CompletableFuture<@NonNull Void> announceNewPunishment(final @NonNull Punishment punishment) {
+    return punishment.getPunishmentType()
+        .getBroadcastMessage(punishment.getReason().isPresent())
+        .map(key -> this.announcePunishment(punishment, key))
+        .orElseGet(() -> CompletableFuture.completedFuture(null));
+  }
+
+  public @NonNull CompletableFuture<@NonNull Void> announceLiftedPunishment(final @NonNull Punishment punishment) {
+    return punishment.getPunishmentType()
+        .getBroadcastLiftMessage()
+        .map(key -> this.announcePunishment(punishment, key))
+        .orElseGet(() -> CompletableFuture.completedFuture(null));
+  }
+
+  public @NonNull IMessage punishmentMessage(final @NonNull MessageKey messageKey,
+      final @NonNull Punishment punishment) {
+    return this.messageFactory.placeholderMessage(messageKey,
+        this.messageFactory.componentComponent(this.createPlaceholders(punishment)));
+  }
+
+  private @NonNull CompletableFuture<@NonNull Void> announcePunishment(final @NonNull Punishment punishment,
+      final @NonNull MessageKey messageKey) {
+    final String permission = punishment.getPunishmentType().getNotificationPermission().orElse(null);
+    return this.punishmentMessage(messageKey, punishment)
+        .component()
+        .thenAccept(component -> {
+          this.banServer.consoleAudience().sendMessage(Identity.nil(), component);
+          for (final IBanAudience audience : this.banServer.onlineAudiences()) {
+            if (permission != null && !audience.hasPermission(permission)) {
+              continue;
+            }
+
+            audience.sendMessage(Identity.nil(), component);
+          }
+        });
+  }
+
+  private @NonNull CompletableFuture<@NonNull IMessageComponent @NonNull []> createPlaceholders(
+      final @NonNull Punishment punishment) {
+    return this.userService.getUser(punishment.getTarget())
+        .thenApply(
+            opt -> opt.orElseThrow(() -> new IllegalArgumentException("Target of punishment cannot be unknown.")))
+        .thenCombine(this.userService.getUser(punishment.getPunisher())
+                .thenApply(opt -> opt
+                    .orElseThrow(() -> new IllegalArgumentException("Punisher of punishment cannot be unknown."))),
+            (target, punisher) -> this.createPlaceholders(punishment, target, punisher))
```

Comment:
This looks cursed. It isn't very clear to see what's going on here with the different levels of nesting. Not sure if this is fixable or not.

Reply:
Not really fixable, no.